pyccel/codegen/printing/ccudacode.py

```diff
@@ -0,0 +1,591 @@
+            information (name, input, output) needed to create the
+            function signature and return a string describing the
+            function.
+            This is not a declaration as the signature does not end
+            with a semi-colon.
+            Parameters
+            ----------
+            expr : FunctionDef
+                The function definition for which a signature is needed.
+            print_arg_names : bool, default : True
+                Indicates whether argument names should be printed.
+            Returns
+            -------
+            str
+                Signature of the function.
+            """
+            if len(expr.results) > 1:
+                self._additional_args.append(expr.results)
+            args = list(expr.arguments)
+            if len(expr.results) == 1:
+                ret_type = self.get_declare_type(expr.results[0])
+            elif len(expr.results) > 1:
+                ret_type = self._print(datatype('int'))
+                args += [FunctionDefArgument(a) for a in expr.results]
+            else:
+                ret_type = self._print(datatype('void'))
+            name = expr.name
+            if not args:
+                arg_code = 'void'
+            else:
+                def get_var_arg(arg, var):
+                    code = "const " * var.is_const
+                    code += self.get_declare_type(var) + ' '
+                    code += arg.name * print_arg_names
+                    return code
+
+                var_list = [a.var for a in args]
+                arg_code_list = [self.function_signature(var, False) if isinstance(var, FunctionAddress)
+                                    else get_var_arg(arg, var) for arg, var in zip(args, var_list)]
+                arg_code = ', '.join(arg_code_list)
+
+            if self._additional_args :
+                self._additional_args.pop()
+
+            extern_word = 'extern "C"'
+            cuda_deco = "__global__" if 'kernel' in expr.decorators else ''
+
+            if isinstance(expr, FunctionAddress):
+                return f'{extern_word} {ret_type} (*{name})({arg_code})'
+            else:
+                return f'{extern_word} {cuda_deco} {ret_type} {name}({arg_code})'
+
+    def _print_Allocate(self, expr):
+        free_code = ''
+        #free the array if its already allocated and checking if its not null if the status is unknown
+        if  (expr.status == 'unknown'):
+            free_code = 'if (%s.shape != NULL)\n' % self._print(expr.variable.name)
+            free_code += "{{\n{}}}\n".format(self._print(Deallocate(expr.variable)))
+        elif  (expr.status == 'allocated'):
+            free_code += self._print(Deallocate(expr.variable))
+        shape = ", ".join(self._print(i) for i in expr.shape)
+        shape_dtype = self.find_in_dtype_registry('int', 8)
+        tmp_shape = self.scope.get_new_name('tmp_shape')
+        dtype = self._print(expr.variable.dtype)
```

Comment:
The `expr.variable` property is used a lot here. We should declare `variable` (see https://github.com/pyccel/pyccel/pull/1357)